rllib/agents/dqn/simple_q.py

```diff
@@ -215,3 +245,61 @@ def execution_plan(workers, config, **kwargs):
         )
 
         return StandardMetricsReporting(train_op, workers, config)
+
+    @ExperimentalAPI
+    def training_iteration(self) -> ResultDict:
+        """Default single iteration logic of an algorithm.
+
+        - Collect on-policy samples (SampleBatches) in parallel using the
+          Trainer's RolloutWorkers (@ray.remote).
+        - Concatenate collected SampleBatches into one train batch.
+        - Note that we may have more than one policy in the multi-agent case:
+          Call the different policies' `learn_on_batch` (simple optimizer) OR
+          `load_batch_into_buffer` + `learn_on_loaded_batch` (multi-GPU
+          optimizer) methods to calculate loss and update the model(s).
+        - Return all collected metrics for the iteration.
+
+        Returns:
+            The results dict from executing the training iteration.
+        """
+        batch_size = self.config["train_batch_size"]
+
+        # Collects SampleBatches in parallel and synchronously
+        # from the Trainer's RolloutWorkers until we hit the
+        # configured `train_batch_size`.
+        # TODO Artur: See what happens if we add to buffer and train as fast
+        #  as we can instead of waiting for full batches
+        sample_batches = []
+        num_env_steps = 0
+        num_agent_steps = 0
+
+        new_sample_batches = synchronous_parallel_sample(self.workers)
```

Comment:
Nice!

Reply:
It works well and took me half an hour to put together. But I am struggling with disentangling other algorithms from the changes I made to get Simple Q to work - as you can see by the number of small commits that I made that constantly make new tests fail.